File: src/bag2video.py
#!/usr/bin/python

import os
import sys
import rospy
import rosbag
import cv2
import cv
import numpy as np
from sensor_msgs.msg import Image as rImage
from sensor_msgs.msg import CompressedImage as cImage


encoding = 'MJPG'
fps = 20.0
outputName = 'output.avi'
isColor = True

if __name__ == '__main__' :
    
    topicName = sys.argv[2]
    if (topicName[-10:] == 'compressed') :
        raw = False
    else :
        raw = True

    videoWriter = None
    mybag = rosbag.Bag (sys.argv[1], 'r')
    countMessage = 0
    topicInfo = mybag.get_type_and_topic_info(topicName)
    numMessage = topicInfo[1][topicName].message_count

    for topic, msg, tm in mybag.read_messages(topics=[topicName]) :
        
        if (raw==False) :
            imgarray = np.fromstring(msg.data, dtype=np.uint8) 
            imageData = cv2.imdecode(imgarray, -1)
            if (len(imageData.shape)==2) :
                isColor = False
            else :
                isColor = True
        else :
            pass
        if (videoWriter is None) :
            videoWriter = cv2.VideoWriter(outputName, cv.CV_FOURCC(
                encoding[0], 
                encoding[1], 
                encoding[2], 
                encoding[3]),
                fps, (imageData.shape[1], imageData.shape[0]), isColor)
            print ("Size: {}x{}".format(imageData.shape[1], imageData.shape[0]))
            m = videoWriter.isOpened() 
        videoWriter.write (imageData)
        countMessage += 1
        if (countMessage == 2000) :
            break
#         print ("{}/{}\r".format(countMessage, numMessage)),
        print (tm)
    
    videoWriter.release()
    print (".. Done")
    
    
    pass

File: src/bag2video.cc
#include <rosbag/bag.h>
#include <rosbag/view.h>
#include <boost/foreach.hpp>
#include <sensor_msgs/CompressedImage.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <cstdlib>


#define foreach BOOST_FOREACH
using namespace std;

string codec ("MJPG");


int main (int argc, char *argv[])
{
	rosbag::Bag mybag;
	mybag.open (argv[1], rosbag::bagmode::Read);
	float fps = atof (argv[3]);

	vector<string> topics;
	topics.push_back (argv[2]);
	rosbag::View bagview (mybag, rosbag::TopicQuery(topics));
	int numMsg = bagview.size();

	cv::VideoWriter vidwriter;
	int countMsg = 0;

	foreach (rosbag::MessageInstance const m, bagview) {
		sensor_msgs::CompressedImage::ConstPtr cimg = m.instantiate<sensor_msgs::CompressedImage>();

		cv::InputArray imgData (cimg->data);
		cv::Mat img = cv::imdecode (imgData, -1);

		// color, resolution?
		if (vidwriter.isOpened() == false) {
			int w = img.cols;
			int h = img.rows;
			bool isColor;
			int typ;
			if (img.type()==CV_8UC1)
				isColor = false;
			else if (img.type()==CV_8UC3)
				isColor = true;
			vidwriter.open ("./output.avi",
				CV_FOURCC(codec[0], codec[1], codec[2], codec[3]),
				fps,img.size(), isColor);
		}

		vidwriter.write (img);
		countMsg++;
		cout << countMsg << '/' << numMsg << "\r";
		if (countMsg==numMsg) cout << endl;
	}

	mybag.close();
}
